### PR TITLE
Jw gawb 1290, 1289 display attributes and attributes of references

### PR DESCRIPTION
**Pull request by jlmwhitt**
- [x] **Submitter**: Include the JIRA issue number in the PR description
- [x] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [x] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * Add notes on what you've tested
- [x] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [x] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [x] **Submitter**: If you're adding new libraries, sign us up to security updates for them
- [x] Tell the tech lead (TL) that the PR exists if they wants to look at it
- [ ] Anoint a lead reviewer (LR). **Assign PR to LR**
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [ ] **TL** sign off
- [ ] **LR** sign off
- [ ] **Product Owner** sign off
- [ ] **Assign to submitter** to finalize
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Squash commits and merge to develop
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

File: src/cljs/org/broadinstitute/firecloud_ui/page/workspace/data/tab.cljs [hunk 1]
```diff
@@ -17,6 +20,40 @@
...
+  (and (map? attr-value)
+       (= (set (keys attr-value)) #{:entityType :entityName})))
+
+(defn- render-list-item [item]
+  (if (is-single-ref? item)
+    (:entityName item)
+    item))
+
+(defn- get-column-name [entity-type]
+  (cond (= (str entity-type) "sample_set") "Sample"
```
Comment (jmthibault79) [inline src/cljs/org/broadinstitute/firecloud_ui/page/workspace/data/tab.cljs]: What knowledge does the UI have of the data model?  If that is available somewhere, we should use it here as well.  That would also give us e.g. Sample -> Participant relationships.

File: src/cljs/org/broadinstitute/firecloud_ui/common/entity_table.cljs [hunk 2]
```diff
@@ -74,8 +74,14 @@
                                                (common/attribute-list? attr-value)
                                                (let [items (map render-list-item (common/attribute-values attr-value))]
                                                  (if (empty? items)
-                                                   "0 items"
-                                                   (str (count items) " items: " (join ", " items))))
+                                                    (cond (= (str selected-entity-type) ":sample_set") "0 samples"
+                                                          (= (str selected-entity-type) ":pair_set") "0 pairs"
+                                                          (= (str selected-entity-type) ":participant_set") "0 participants"
+                                                          :else  "0 items")
+                                                    (cond (= (str selected-entity-type) ":sample_set") (str (count items) " samples")
```
Comment (jmthibault79) [inline src/cljs/org/broadinstitute/firecloud_ui/common/entity_table.cljs]: After this change, we don't list the samples here.  Is that what we want?   I can see arguments for either way.  I just didn't see this change in the story.

File: src/cljs/org/broadinstitute/firecloud_ui/page/workspace/data/tab.cljs (hunk 1 above)
Comment (jlmwhitt) [reply src/cljs/org/broadinstitute/firecloud_ui/page/workspace/data/tab.cljs]: I have updated so that the entity references in the main table are linked

**Comment by jmthibault79** [comment]
*** NOTE ***

I attempted to clean up this branch after some merge conflicts.  It is VERY LIKELY that some code that doesn't belong here has come along for the ride.  Please review this carefully.

See you in January!

File: src/cljs/org/broadinstitute/firecloud_ui/page/workspace/data/tab.cljs [hunk 3]
```diff
@@ -17,6 +20,41 @@
     [org.broadinstitute.firecloud-ui.utils :as u]
     ))
 
+;; for attributes referring to a single other entity
+(defn- is-single-ref? [attr-value]
```
Comment (dmohs) [inline src/cljs/org/broadinstitute/firecloud_ui/page/workspace/data/tab.cljs]: This is duplicated in `entity_table.cljs` so should be pulled out into a common location.

File: src/cljs/org/broadinstitute/firecloud_ui/common/entity_table.cljs [hunk 4]
```diff
@@ -70,12 +70,23 @@
                                            :content-renderer
                                            (fn [attr-value]
                                              (cond
-                                               (is-single-ref? attr-value) (:entityName attr-value)
+                                               (is-single-ref? attr-value)
+                                               (if (:linked-entity-renderer props)
```
Comment (dmohs) [inline src/cljs/org/broadinstitute/firecloud_ui/common/entity_table.cljs]:
```clojure
(if-let [r (:linked-entity-renderer props)]
  (r attr-value)
  (:entityName attr-value))
```

File: src/cljs/org/broadinstitute/firecloud_ui/page/workspace/data/tab.cljs [hunk 5]
```diff
@@ -58,12 +96,98 @@
...
+                          :always-sort? true
+                          :header-row-style {:borderBottom (str "2px solid " (:line-default style/colors))
+                                             :backgroundColor "white" :color "black" :fontWeight "bold"}
+                          :empty-message (if (= item-column-name "Entity") (str "No Entity Attributes defined")
+                                                                           (str "No " item-column-name "s defined"))
+                          :columns (if (is-entity-set? entity-type) setColumns singleColumns)
+                          :data (seq attributes)
+                          :->row
+                          (fn [x]
+                            (if (map? x)   ;;map is a sample_set or pair_set or participant_set
```
Comment (dmohs) [inline src/cljs/org/broadinstitute/firecloud_ui/page/workspace/data/tab.cljs]: Nit: The convention is usually single-semicolon comments for those that appear on the same line as code, double-semicolon for comments that are alone on a line.

File: src/cljs/org/broadinstitute/firecloud_ui/page/workspace/data/tab.cljs [hunk 6]
```diff
@@ -74,43 +198,82 @@
...
+                                          })))}])
           :else
-          [:div {:style {:textAlign "center"}} [comps/Spinner {:text "Checking workspace..."}]])]))
+          [:div {:style {:textAlign "center"}} [comps/Spinner {:text "Checking workspace..."}]])
+        (let [workspaceId (:workspace-id props)
+              entityType (:current-entity-type @state)
+              entityName (:selected-entity @state)
+              attributes (:attr-list @state)]
+          [EntityAttributes {:workspace-id workspaceId :entity-type entityType :entity-name entityName
+                             :attr-list attributes :mainState state}])]
```
Comment (dmohs) [inline src/cljs/org/broadinstitute/firecloud_ui/page/workspace/data/tab.cljs]: `:mainState state`

**Review by dmohs — CHANGES_REQUESTED** [review]
`EntityAttributes` is swapping the state of its parent. I know it's not your fault, but we gotta fix it.

**Review by dmohs — APPROVED** [review]
Thanks for taking the time to clean this up.

**Comment by zarsky-broad** [comment]
(This last commit done w/ @dmohs, who says he approves, so I'm totally following procedure if anybody asks, which they likely won't, but it's the last day before the holiday break so I'm writing this here anyway; hello, posterity)

**Pull request merged**